core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java

```diff
@@ -132,59 +146,154 @@ public UnicastZenPing(Settings settings, ThreadPool threadPool, TransportService
+     * in parallel using the generic thread pool from the specified thread pool up to the specified resolve timeout.
      *
-     * @param host the host to resolve
-     * @param limitPortCounts the number of ports to resolve (should be 1 for non-local transport)
+     * @param threadPool       the thread pool used to parallelize hostname lookups
+     * @param logger           logger used for logging messages regarding hostname lookups
+     * @param hosts            the hosts to resolve
+     * @param limitPortCounts  the number of ports to resolve (should be 1 for non-local transport)
      * @param transportService the transport service
-     * @param idGenerator the generator to supply unique ids for each discovery node
+     * @param idGenerator      the generator to supply unique ids for each discovery node
+     * @param resolveTimeout   the timeout before returning from hostname lookups
      * @return a list of discovery nodes with resolved transport addresses
      */
-    public static List<DiscoveryNode> resolveDiscoveryNodes(final String host, final int limitPortCounts,
-                                                            final TransportService transportService, final Supplier<String> idGenerator) {
-        List<DiscoveryNode> discoveryNodes = new ArrayList<>();
-        try {
-            TransportAddress[] addresses = transportService.addressesFromString(host, limitPortCounts);
-            for (TransportAddress address : addresses) {
-                discoveryNodes.add(new DiscoveryNode(idGenerator.get(), address, emptyMap(), emptySet(),
-                                                        Version.CURRENT.minimumCompatibilityVersion()));
+    public static List<DiscoveryNode> resolveDiscoveryNodes(
+        final ThreadPool threadPool,
+        final Logger logger,
+        final List<String> hosts,
+        final int limitPortCounts,
+        final TransportService transportService,
+        final Supplier<String> idGenerator,
+        final TimeValue resolveTimeout) throws InterruptedException {
+        Objects.requireNonNull(threadPool);
+        Objects.requireNonNull(logger);
+        Objects.requireNonNull(hosts);
+        Objects.requireNonNull(transportService);
+        Objects.requireNonNull(idGenerator);
+        Objects.requireNonNull(resolveTimeout);
+        if (resolveTimeout.nanos() <= 0) {
+            throw new IllegalArgumentException("resolve timeout must be positive but was [" + resolveTimeout + "]");
+        }
+        // create tasks to submit to the executor service; we will wait up to resolveTimeout for these tasks to complete
+        final List<Callable<ResolvedHostname>> callables =
+            hosts.stream().map(hn -> lookup(hn, transportService, limitPortCounts)).collect(Collectors.toList());
+        final List<Future<ResolvedHostname>> futures =
+            threadPool.generic().invokeAll(callables, resolveTimeout.nanos(), TimeUnit.NANOSECONDS);
```

Comment:
The reason I went with this approach is since this method is also used by the file-based unicast hosts provider so it would need its own executor and we do not have a close method for plugins to allow this executor be shutdown.